test/generate/test_generate.py

```diff
@@ -0,0 +1,285 @@
+
+from torchtitan import utils
+
+from torchtitan.config_manager import JobConfig
+from torchtitan.datasets import build_tokenizer
+from torchtitan.logging import init_logger, logger
+from torchtitan.metrics import build_gpu_memory_monitor
+from torchtitan.models import model_name_to_cls, model_name_to_tokenizer, models_config
+from torchtitan.parallelisms import models_parallelize_fns, ParallelDims
+
+# support running w/o installing as package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from generate.generation import apply_torchchat_tp, generate
+
+
+@record
+def example_generate(
+    config_path: str,
+    checkpoint_path: str,
+    prompt: str,
+    *,
+    temperature: float = 1.0,
+    max_new_tokens: int = 32,
+    batch_size: int = 1,
+    top_k: Optional[int] = None,
+    seed: Optional[int] = None,
+):
+    init_logger()
+    color = utils.Color
+
+    # Load configuration from toml file
+    config = JobConfig()
+    config.parse_args([f"--job.config_file={config_path}"])
+    config._validate_config()
+
+    utils.set_determinism(seed)
+
+    if seed is None:
+        logger.info("Deterministic sampling off")
+    else:
+        logger.info(f"Deterministic sampling on. Using seed: {seed}")
+
+    world_size = int(os.environ.get("WORLD_SIZE", 1))
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    device = torch.device(f"cuda:{local_rank}")
+    torch.cuda.set_device(device)
+    gpu_memory_monitor = build_gpu_memory_monitor()
+
+    model_name = config.model.name
+
+    # Init distributed env
+    if world_size > 1:
+        utils.init_distributed(config)
+        parallel_dims = ParallelDims(
+            dp_replicate=1,
+            dp_shard=-1,
+            cp=1,
+            tp=world_size,
+            pp=1,
+            world_size=world_size,
+            enable_loss_parallel=False,
+        )
+        # Build world mesh for parallelism
+        world_mesh = parallel_dims.build_mesh(device_type="cuda")
+
+    logger.info(f"World Size: {world_size}, Local Rank: {local_rank} on {device}")
+
+    # Tokenizer setup
+    tokenizer = build_tokenizer(
+        model_name_to_tokenizer[model_name], config.model.tokenizer_path
+    )
+
+    model_config = models_config[model_name][config.model.flavor]
+    model_config.norm_type = config.model.norm_type
+    model_config.max_seq_len = config.training.seq_len
+    model_config.vocab_size = tokenizer.n_words
+
+    model_cls = model_name_to_cls[model_name]
+    init_device = "meta" if world_size > 1 else device
+    with torch.device(init_device):
+        logger.info(f"Init model on init_device: {init_device}")
+        model = model_cls.from_model_args(model_config)
+
+    if world_size > 1:
+        use_torchchat_tp = False
+        if use_torchchat_tp:
+            apply_torchchat_tp(model, world_mesh["tp"])  # Working
+        else:
+            models_parallelize_fns[model_name](model, world_mesh, parallel_dims, config)
```

Comment:
I've figured out what's the bug with odd length sequences.

The issue was caused by "sharded sequences on different ranks forgot the `seq_len`, and made wrong guess based on the local sharded length". cc: @wz337 

To mitigate, we can either disable Sequence Parallel (the same as what torchchat does -- anyway SP is not necessary here), or land my change at https://github.com/pytorch/torchtitan/pull/686.